### ISSUE TITLE
[explicit-function-return-type] infer the type from the return value for bodyless arrow functions

### ISSUE THREAD

**Issue by dandv** [typescript-eslint/typescript-eslint#1731]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/explicit-function-return-type": "warn"
  }
}
```

```TS
function foo(param1: number, ..., callback = () => undefined) { ... }
```

**Expected Result**
No warning

**Actual Result**
> `ESLint: Missing return type on function.(@typescript-eslint/explicit-function-return-type)`

**Additional Info**
Passing a no-op as the default value for a callback saves checking for that callback being defined before calling it.

Annotating the return type to `void` only serves the TypeScript parser, not the developer.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.19.0` |
| `@typescript-eslint/parser`        | `2.19.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `13.10.0` |

**Comment by bradzacher** [comment]
> Passing a no-op as the default value for a callback is pretty common

An optional call expression saves an explicit check:
```ts
function foo(param1: number, ..., callback?: SomeFunctionType) {
  callback?.(someArg);
}
```

I generally advise people against adding functions as default argument values because default values are re-evaluated every function call, so each call to `foo` causes a new empty function to be created and allocated if no callback is provided.

There are various options which might help you out here; `allowExpressions` will not check any arrow functions

I would have suggested `allowTypedFunctionExpressions`, but right now it doesn't yet support argument assignment patterns with type annotations.
It could be enhanced to check this case, if you're happy to submit a small PR to do so.

**Comment by dandv** [comment]
Good point about not passing a function as the default argument.

Perhaps a more common situation is one-liners that return strings:

    const formatter = value => `some formatting applied to ${value}`;

**Comment by NonPolynomial** [comment]
> I generally advise people against adding functions as default argument values because default values are re-evaluated every function call, so each call to `foo` causes a new empty function to be created and allocated if no callback is provided.

Thats why you should use `Function.prototype` as default, because its a global no-op function in JS

**Comment by kirkwaiblinger** [comment]
I'm thinking this is basically solved with https://github.com/typescript-eslint/typescript-eslint/issues/8950 being completed, so I'm closing for now. 

Someone speak up, though, if this issue is still needed
